Initial openstack support

TheRealFalcon opened the pull request:
At this point I think the basic functionality is usable and reviewable.

Everything in the base_api.py test works, so the basics all work. I intentionally didn't add an example since all clouds should share the same base API right now. We should really only need one example unless we want to showcase specialized use cases of a specific cloud.

TheRealFalcon commented on pycloudlib/openstack/cloud.py:
```diff
@@ -0,0 +1,159 @@
+
+    _type = 'openstack'
+
+    def __init__(self, tag, network, timestamp_suffix=True):
```
I think we have no choice but to require a network parameter. Pycloudlib can't auto-determine the network to use because there could be many defined for any particular openstack deployment and randomly choosing one seems like a bad idea.

TheRealFalcon commented on pycloudlib/openstack/cloud.py:
```diff
@@ -0,0 +1,159 @@
+        self._log.warning('Released images NOT supported for openstack')
+        return self.daily_image(release)
+
+    def daily_image(self, release, **kwargs):
```
I initially tried to use simplestreams here, but realized that many openstack deployments (including ours) are probably out of date with the latest images.

OddBloke replied:
Yeah, that's definitely an issue we'll hit with OpenStack deployments. It's also entirely possible, of course, that there are no Ubuntu images available in any given deployment.

I'm concerned that this heuristic could cause some confusing test behaviours, both between platforms (i.e. a current EC2 image vs a stale OpenStack image) and between OpenStack deployments (most likely one which is running glance-simplestreams-sync vs one which is not). We could also very easily end up launching non-base non-official images without noticing.

I wonder if instead of trying to determine this automatically, we should accept that we can't reliably launch the images that pycloudlib usually does and require OpenStack consumers to provide an explicit image ID in all cases? My thinking is that the image ID to use is deployment-specific, so is best determined by the consumers who know what OpenStack deployment they're running against.

(IIRC, what we did in cloud-test-framework is have an OpenStack cloud class which holds the cross-deployment behaviour and logic, and then subclasses of that which define the deployment-specific behaviour: ScalingStack and then the OpenStack public clouds we supported. Someone trying to use "generic OpenStack" would have to fill in the unknowns to proceed, but we still had a shared understanding of how to operate against shared deployments stored in the codebase.)

A couple of other thoughts: I believe that glance-simplestreams-sync does include additional metadata against the images, which we might be able to use to make this heuristic less likely to hit false positives (i.e. less likely to return non-official Ubuntu images); we might even be able to use it to cross-reference streams (i.e. gather a list of all fingerprints that are valid for $series according to streams and then find the most recent corresponding image). We could also consider using streams to populate the target deployment with an image ourselves: that's probably the only way we can be sure that we'll have an official Ubuntu image on every OpenStack deployment we encounter.

TheRealFalcon commented on pycloudlib/openstack/cloud.py:
```diff
@@ -0,0 +1,159 @@
+                    passed_key=public_key_content
+                )
+            )
+        # TODO: Figure out if I want to incorpoate this...
```
If anybody has strong opinions about this, please chime in.

TheRealFalcon commented:
I recently noticed that on the CLI, I don't need to provide the network information like I do through the API. I'm curious how the CLI knows which network to create the instance on and if we could use that same knowledge to acquire the network to use when using the API.

TheRealFalcon commented:
I'm also going on the assumption that we want a floating IP associated with instances that we launch. Obviously not every openstack deployment in the wild will require that, but I don't think that's something we need to worry about yet.

TheRealFalcon commented on docs/clouds/openstack.md:
```diff
@@ -0,0 +1,21 @@
+# Openstack
```
Added some documentation here, but not sure how I'm supposed to generate it. I tried the makefile in the docs directory, but the output looks very not right.

OddBloke replied:
https://github.com/canonical/pycloudlib/pull/125

OddBloke commented on pycloudlib/openstack/tests/test_cloud.py:
```diff
@@ -0,0 +1,40 @@
+@mock.patch('pycloudlib.key.KeyPair.public_key_content',
+            new_callable=mock.PropertyMock,
+            return_value='pretend public key')
+@mock.patch('openstack.connect', spec=Connection)
```
I'm not sure this `spec` is doing what you want here:

```py
In [1]: import unittest.mock as mock
In [3]:  from openstack.connection import Connection
In [4]: p = mock.patch("openstack.connect", spec=Connection)
In [5]: m = p.start()
In [6]: m()  # this is what `openstack.connect()` will return
Out[6]: <MagicMock name='connect()' id='140181990886512'>
In [7]: m  # and I think you actually want it to return this
Out[7]: <openstack.connection.Connection at 0x7f7ea9c62fa0>
```

OddBloke commented on pycloudlib/openstack/tests/test_cloud.py:
```diff
@@ -0,0 +1,40 @@
+class TestOpenstackKeypair:
+    """Tests covering _get_openstack_keypair."""
+
+    def test_keypair_doesnt_exist(self, m_openstack, m_keypair):
```
```suggestion
    def test_keypair_doesnt_exist(self, m_openstack, _m_keypair):
```

OddBloke commented on pycloudlib/openstack/tests/test_cloud.py:
```diff
@@ -0,0 +1,40 @@
+        cloud._get_openstack_keypair()
+        assert 1 == cloud.conn.create_keypair.call_count
+
+    def test_keypairs_match(self, m_openstack, m_keypair):
```
`m_keypair` threw me off here: I assumed that would be a mock of `pycloudlib.key.KeyPair` (making this test very confusing :p). Perhaps `m_public_key_content`? Still a little confusing, given it's a property mock (whereas you'd never call `public_key_content` of course).

OddBloke reviewed (commented):
Thanks for this work, James! By and large, this looks good. I have some thoughts inline.

TheRealFalcon commented on pycloudlib/openstack/cloud.py:
```diff
@@ -0,0 +1,189 @@
+        instance = self.conn.compute.create_server(
+            name=self.tag,
+            image_id=image_id,
+            flavor_id=self.conn.compute.find_flavor(instance_type).id,
```
Ah, I didn't realize that about flavors. I'll update it accordingly.

TheRealFalcon commented on pycloudlib/openstack/instance.py:
```diff
@@ -0,0 +1,125 @@
+    def console_log(self):
+        """Return the instance console log."""
+        # Returning nothing for me but might work in other contexts?
+        self.conn.compute.get_server_console_output(self.server)
```
:facepalm: I guess that explains why it was returning nothing!

TheRealFalcon commented:
Addressed the code comments. Still need to address the test comments. @OddBloke , I addressed the floating IP issue a little differently than we initially discussed, so you might want to take a look.

TheRealFalcon commented:
Updated the test contents, but now I have new floating ip tests to write :wink:

TheRealFalcon commented:
Added new tests. Should be ready for re-review.

OddBloke commented on pycloudlib/openstack/instance.py:
```diff
@@ -0,0 +1,151 @@
+        self.delete_floating_ip = False
+        self.floating_ip = self._get_existing_floating_ip()
+        if self.floating_ip is None:
+            self._create_and_attach_floating_id()
```
A nit, but `self.floating_ip = self._create_and_attach_floating_id()` feels a little easier to follow; it would give `_get_...` and `_create_...` a (more) similar API and means that `self.floating_ip` is only set in one place.

OddBloke commented on pycloudlib/openstack/tests/test_instance.py:
```diff
@@ -0,0 +1,66 @@
+
+# While the actual results from openstack will have a lot more values
+# in the dict, 'floating_ip_address' is all we care about
+NETWORK_IPS = [
+    {'floating_ip_address': '10.0.0.4'},
+    {'floating_ip_address': '10.0.0.3'},
+]
```
Another nit, but if the code under test incorrectly did the equivalent of `chain(*[d.values() for d in NETWORK_IPS])`, we wouldn't catch that with this input.

OddBloke commented on pycloudlib/openstack/tests/test_instance.py:
```diff
@@ -0,0 +1,66 @@
+    def test_existing_floating_ip(self, m_create):
+        """Test that if a server has an existing floating IP, we use it."""
+        m_connection = mock.Mock()
+        m_server = m_connection.compute.get_server.return_value = mock.Mock()
```
A nit; I don't think we need this final assignment:

```suggestion
        m_server = m_connection.compute.get_server.return_value
```

```py
In [6]: from unittest import mock                                                                                                                                                                                                                                                         

In [7]: m_connection = mock.Mock()                                                                                                                                                                                                                                                        

In [8]: m_server = m_connection.compute.get_server.return_value                                                                                                                                                                                                                           

In [9]: m_server                                                                                                                                                                                                                                                                          
Out[9]: <Mock name='mock.compute.get_server()' id='140123963984144'>

In [10]: m_server = m_connection.compute.get_server.return_value = mock.Mock()                                                                                                                                                                                                            

In [11]: m_server                                                                                                                                                                                                                                                                         
Out[11]: <Mock name='mock.compute.get_server()' id='140123963995232'>
```

OddBloke reviewed (commented):
Floating IP stuff LGTM, I've left a few inline comments, all of which are nits. I'm going to run some tests locally before approving but, assuming that's all good, this will be good to land.

OddBloke reviewed (approved):
Thanks for all the work!
